chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -264,7 +260,7 @@ where
         from: u64,
         to: u64,
         filter: EthGetLogsFilter,
-    ) -> impl Future<Item = Vec<Log>, Error = Error> {
+    ) -> Pin<Box<dyn std::future::Future<Output = Result<Vec<Log>, Error>> + Send>> {
```

Comment:
I put a name in the prelude to make this common pattern terse. Prefer `TryDynFuture<'static, Vec<Log>>`. 

For bonus points, you can do `TryDynFuture<'a, Vec<Log>>` to allow borrowing from self and other arguments in the future - eliminating both the existing pesky clones like the `let eth = self.clone()` on line 280, as well as the freshly introduced clones on 291 - 294. I verified that doing so compiles and removes all the clones from the method except for the Arc (which could also be removed if callees are also modified to take references)

The idiomatic thing to do is to return `TryDynFuture<'_, Vec<Log>>` and take _all_ non-`Copy` types by reference when migrating to async-await. This is just a suggestion though because it means fixing the caller as well.

Reply:
Oh nice I'll use `TryDynFuture`.

Did you manage to also fix the caller (`logs_in_block_range`) when trying that change? The resulting lifetime error is pretty nasty.

Reply:
Instead of fixing it I cut it off with a `todo()! `just to test changes to the first function in isolation.

The way I've been fixing these is just to migrate the use of references up the stack until the compiler stops complaining. Sometimes this means all the way up to main. It's a bit of a pain, but each function migrated is progress so it'll get easier with time as it's fewer and fewer steps up the callstack before finding something that's been migrated already. Sometimes you luck out though and the compiler is just happy.

Reply:
I fixed a lot of these in this PR - https://github.com/graphprotocol/graph-node/pull/1498
So if you get stuck on a lifetime error let's screen share.

Reply:
While I am curious to know the solution to this particular lifetime puzzle, I wouldn't want to bubble up lifetimes through the calling functions.

All these clones are cheap, and insignificant compared to a network call, so this would be more a discussion on code style. On that, `'static` things have less of an annotational burden and most importantly are easier to work with, lifetimes would will generate more lifetime puzzles up the stack, which costs development time now and in the future.

Reply:
Oops, my suggestion to use `DynTryFuture` may be misguided here (though it's not wrong).

Better would be to return some kind of `impl Future`. `Box<dyn Future<..>>` is a hack to get around the fact that GAT is required to return `impl Future` from traits. Since traits are not in play here, the `Box` is unnecessary and the quickest way to get the `impl Future` is just to use the `async` keyword.

Reply:
I strongly disagree with the points for not using lifetimes.

> All these clones are cheap

Didn't you just fix a bug that was eating up gigabytes because of a seemingly innocent `clone()`?. In the previous PR I just linked to there were a bunch of clones that were removed which were not cheap as well. The problem is that when the use of clone is ubiquitous it is easy to miss performance problems because the clones become noise.

(That gives me an idea for a `trait CheapClone` which would be implemented for things like `Arc`, but not `Vec` - making the performance story more self-documenting)

Performance aside, a clone is semantically different than a shared reference. A clone gives you owned mutable data. A shared reference keeps things in sync.

The way that lifetimes work, with bubbling up the stack being the solution to the problem we are encouraged to use a consistent style across the codebase. Things become more difficult by trying to intermingle the two styles (by reference vs &'static), async, not async.

> On that, 'static things have less of an annotational burden and most importantly are easier to work with

They are more familiar to you perhaps but are not easier to work with. The style of this codebase is more of an outlier in this regard. Once you start working with mutable data references are the only reasonable option. Don't be afraid of references. Working with them will make you a much better Rust developer.

Reply:
The first thing I tried was making `log_stream` an `async fn`, but that resulted in the most inscrutable Rust error I've seen in while, it said "one type is more general than the other" :man_shrugging:.

I've introduced a `CheapClone` trait and used it in all these clones. The `filter` was being cloned around a bunch so I put it in an `Arc`. It's a vec of event signatures and contract addresses, I even [benchmarked]( https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=78afed01372536d38800a541800189be) it out of curiosity to see the impact of cloning a structure like that.

On trying to use lifetimes on everything, if we had more async/await on the path between the task being spawned and this function I might feel more adventurous to try it, but that refactor would be disproportional to the size of the change in this PR.

Reply:
Ooh, I got that "more general" error once when learning Rust - for a completely different reason. This time it might have been [needing to add async move](https://github.com/rust-lang/rust/issues/64650).

Glad to see the CheapClone trait working out. The one thing I'd do there is add an `#[inline]` attribute.

In any case, I'm happy with the code as-is. These and previous comments are all just possible areas for improvement so let's get this merged.